src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Automation/VsBuildManager.cs

```diff
@@ -0,0 +1,132 @@
+        public event _dispBuildManagerEvents_DesignTimeOutputDeletedEventHandler DesignTimeOutputDeleted;
+
+        /// <summary>
+        /// Occurs when a design time output moniker is dirty
+        /// </summary>
+        public event _dispBuildManagerEvents_DesignTimeOutputDirtyEventHandler DesignTimeOutputDirty;
+
+        #endregion
+
+        /// <summary>
+        /// Gets the project of which the selected item is a part.
+        /// </summary>
+        public EnvDTE.Project ContainingProject => Project.FirstOrDefault()?.Value.Project;
+
+        /// <summary>
+        /// Gets the top-level extensibility object.
+        /// </summary>
+        public EnvDTE.DTE DTE => Project.FirstOrDefault()?.Value.DTE;
+
+        /// <summary>
+        /// Gets the immediate parent object of a given object.
+        /// </summary>
+        public object Parent => Project.FirstOrDefault()?.Value;
+
+        /// <summary>
+        /// Gets the temporary portable executable (PE) monikers for a project.
+        /// </summary>
+        public object DesignTimeOutputMonikers
+        {
+            get
+            {
+                return _unconfiguredProjectServices.ThreadingService.ExecuteSynchronously(_tempPEManager.GetDesignTimeOutputFilenamesAsync);
+            }
+        }
+
+        #region BuildManager Members
+
+        /// <summary>
+        /// Builds a temporary portable executable (PE) and returns its description in an XML string.
+        /// </summary>
+        public string BuildDesignTimeOutput(string bstrOutputMoniker)
+        {
+            if (bstrOutputMoniker == null) return null;
+
+            return _unconfiguredProjectServices.ThreadingService.ExecuteSynchronously(() =>
+            {
+                return _tempPEManager.GetTempPEBlobAsync(bstrOutputMoniker);
+            });
+        }
+
+        #endregion
+
+        #region IEventSource<_dispBuildManagerEvents> Members
+
+        void IEventSource<_dispBuildManagerEvents>.OnSinkAdded(_dispBuildManagerEvents sink)
+        {
+            DesignTimeOutputDeleted += new _dispBuildManagerEvents_DesignTimeOutputDeletedEventHandler(sink.DesignTimeOutputDeleted);
+            DesignTimeOutputDirty += new _dispBuildManagerEvents_DesignTimeOutputDirtyEventHandler(sink.DesignTimeOutputDirty);
+        }
+
+        void IEventSource<_dispBuildManagerEvents>.OnSinkRemoved(_dispBuildManagerEvents sink)
+        {
+            DesignTimeOutputDeleted -= new _dispBuildManagerEvents_DesignTimeOutputDeletedEventHandler(sink.DesignTimeOutputDeleted);
+            DesignTimeOutputDirty -= new _dispBuildManagerEvents_DesignTimeOutputDirtyEventHandler(sink.DesignTimeOutputDirty);
+        }
+
+        #endregion
+
+        /// <summary>
+        /// Occurs when a design time output moniker is deleted.
+        /// </summary>
+        internal virtual void OnDesignTimeOutputDeleted(string outputMoniker)
+        {
+            DesignTimeOutputDeleted?.Invoke(outputMoniker);
```

Comment:
Is this what CPS does? In legacy, any consumers of BuildManager will be RPC marshalled to the UI thread, and the these events will only ever being fired on UI thread. Please assert/check that we're on UI thread.

Reply:
Yes, this is straight from CPS too. The ReferencesHostBridge uses `HResult.Invoke` to fire its events.

Once I change ApplyAsync to be on the UI thread this will be too though.

Reply:
Please assert that.